src/source/load_vector_tile.ts

```diff
@@ -38,46 +62,143 @@ export class DedupedRequest {
+                        callback,
+                        fromQueue: true
+                    }).cancel;
+                }
+            }
+        };
 
         if (entry.result) {
             const [err, result] = entry.result;
-            if (this.scheduler) {
-                this.scheduler.add(() => {
-                    callback(err, result);
-                }, metadata);
-            } else {
-                callback(err, result);
-            }
-            return () => {};
+            this.addToSchedulerOrCallDirectly({
+                callback,
+                metadata,
+                err,
+                result,
+            });
+            return {cancel: () => {}};
         }
 
-        entry.callbacks.push(callback);
+        entry.callbacks.add(callback);
 
-        if (!entry.cancel) {
-            entry.cancel = request((err, result) => {
+        const inQueue = requestQueue.has(key);
+        if ((!entry.cancel && !inQueue) || fromQueue) {
+            // Lack of attached cancel handler means this is the first request for this resource
+            if (numRequests >= 50) {
+                const queued = {
+                    key,
+                    metadata,
+                    requestFunc,
+                    callback,
+                    cancelled: false,
+                    cancel() {},
+                };
+                const cancelFunc = () => {
+                    queued.cancelled = true;
+                    removeCallbackFromEntry({
+                        key,
+                        requestCallback: callback,
+                    });
+                };
+                queued.cancel = cancelFunc;
+                requestQueue.set(key, queued);
+                return queued;
+            }
+            numRequests++;
+
+            const actualRequestCancel = requestFunc((err, result) => {
                 entry.result = [err, result];
-                for (const cb of entry.callbacks) {
-                    if (this.scheduler) {
-                        this.scheduler.add(() => {
-                            cb(err, result);
-                        }, metadata);
-                    } else {
-                        cb(err, result);
-                    }
-                }
-                setTimeout(() => delete this.entries[key], 1000 * 3);
+
+                // Notable difference here compared to previous deduper, no longer iterating through callbacks stored on the entry
+                // Due to intermittent errors thrown when duplicate arrayBuffers get added to the scheduling
+                this.addToSchedulerOrCallDirectly({
+                    callback,
+                    metadata,
+                    err,
+                    result,
+                });
+
+                filterQueue(key);
+                advanceImageRequestQueue();
+
+                setTimeout(() => {
+                    delete this.entries[key];
+                }, 1000 * 3);
```

Comment:
just wondering why 1000 * 3 is used here as timeout instead of 3000?

Reply:
Convention in other parts of the codebase, but no reason outside of that 🙂